### PR TITLE
integration tests: fix docker-env test in powershell

### PR DESCRIPTION
**Pull request by medyagh**
This PR reduces the windows failure tests from 8 to 7.
closes https://github.com/kubernetes/minikube/issues/8180

After multiple tries found out golang os/exec fails running properly in powershell and it needs extra care to run powershell.

#### special thanks to the person who posted this in gist. thank you @coolbrg
https://gist.github.com/coolbrg/d1854cc771025efb4a30197820c2c612

### CODE REVIEW

**Comment by tstromberg** [comment]
Please ensure that the PR title is a declaration of what is being changed by this PR. It should make sense when displayed in the release notes.

File: test/integration/fn_mount_cmd.go [hunk 1]
```diff
@@ -62,10 +62,10 @@ func validateMountCmd(ctx context.Context, t *testing.T, profile string) {
 		t.Fatalf("Unexpected error while creating tempDir: %v", err)
 	}
 
-	ctx, cancel := context.WithTimeout(ctx, Minutes(10))
```
Comment (tstromberg) [inline test/integration/fn_mount_cmd.go:65]: Move unrelated context changes into another PR. There is too much going on in this PR to decipher why this change is being done.

File: test/integration/functional_test.go [hunk 2]
```diff
@@ -155,23 +155,43 @@ func validateNodeLabels(ctx context.Context, t *testing.T, profile string) {
...
 
-	mctx, cancel := context.WithTimeout(ctx, Seconds(13))
+	mctx, cancel := context.WithTimeout(ctx, Seconds(30))
 	defer cancel()
-	// we should be able to get minikube status with a bash which evaled docker-env
-	c := exec.CommandContext(mctx, "/bin/bash", "-c", "eval $("+Target()+" -p "+profile+" docker-env) && "+Target()+" status -p "+profile)
-	rr, err := Run(t, c)
+	var rr *RunResult
+	var err error
+	if runtime.GOOS == "windows" { // golang exec powershell needs some tricks !
```
Comment (tstromberg) [inline test/integration/functional_test.go:162]: This comment is not helpful in this context. Remove it.

File: test/integration/functional_test.go [hunk 3]
```diff
@@ -155,23 +155,43 @@ func validateNodeLabels(ctx context.Context, t *testing.T, profile string) {
...
+		c := exec.CommandContext(mctx, Target()+" -p "+profile+" docker-env | Invoke-Expression ;"+Target()+" status -p "+profile)
+		rr, err = Run(t, c, true) // golang exec powershell needs some tricks !
+	} else {
+		c := exec.CommandContext(mctx, "/bin/bash", "-c", "eval $("+Target()+" -p "+profile+" docker-env) && "+Target()+" status -p "+profile)
+		// we should be able to get minikube status with a bash which evaled docker-env
+		rr, err = Run(t, c)
+	}
+	if ctx.Err() == context.DeadlineExceeded {
```
Comment (tstromberg) [inline test/integration/functional_test.go:170]: I think you are checking the error on the wrong context here.

File: test/integration/functional_test.go [hunk 4]
```diff
@@ -155,23 +155,43 @@ func validateNodeLabels(ctx context.Context, t *testing.T, profile string) {
...
+		t.Errorf("failed to run the command in 30 seconds. exceeded 30s timeout. %s", rr.Command())
```
Comment (tstromberg) [inline test/integration/functional_test.go:171]: Remove the 30s string - it's going to almost certainly go stale.

File: test/integration/functional_test.go [hunk 5]
```diff
@@ -816,8 +836,34 @@ func validateSSHCmd(ctx context.Context, t *testing.T, profile string) {
 	if NoneDriver() {
 		t.Skipf("skipping: ssh unsupported by none")
 	}
-	want := "hello\n"
-	rr, err := Run(t, exec.CommandContext(ctx, Target(), "-p", profile, "ssh", fmt.Sprintf("echo hello")))
+	want := "/home/docker\n"
```
Comment (tstromberg) [inline test/integration/functional_test.go:839]: I don't understand the point of this new test. Why is /home/docker a critical testing value? The echo command seemed a better indicator of badness, to be honest.

File: test/integration/functional_test.go [hunk 6]
```diff
@@ -816,8 +836,34 @@ func validateSSHCmd(ctx context.Context, t *testing.T, profile string) {
...
+
+	if err != nil {
+		t.Errorf("failed to run an ssh command. args %q : %v", rr.Command(), err)
+	}
+
+	if rr.Stdout.String() != want {
+		t.Errorf("expected minikube ssh command output to be -%q- but got *%q*. args %q", want, rr.Stdout.String(), rr.Command())
+	}
+
+	// testing hostname as well
```
Comment (tstromberg) [inline test/integration/functional_test.go:855]: This comment is missing the "why" aspect. Why do we care that the hostname matches the profile exactly?

File: test/integration/helpers.go [hunk 7]
```diff
@@ -86,22 +87,44 @@ func (rr RunResult) Output() string {
...
+func Run(t *testing.T, cmd *exec.Cmd, powershell ...bool) (*RunResult, error) {
 	t.Helper()
-	rr := &RunResult{Args: cmd.Args}
+	runInPowershell := false
+	if len(powershell) > 0 {
+		runInPowershell = powershell[0]
+	}
+
+	var newCmd *exec.Cmd
+	if runInPowershell {
```
Comment (tstromberg) [inline test/integration/helpers.go:98]: Particularly since this is a test, please implement this in an opaque nature rather than hiding it into a boolean.

For now, just pass the cmd in configured properly for the few powershell cases we have.

File: test/integration/helpers.go [hunk 8]
```diff
@@ -86,22 +87,44 @@ func (rr RunResult) Output() string {
...
+		psBin, err := exec.LookPath("powershell.exe")
```
Comment (tstromberg) [inline test/integration/helpers.go:99]: This isn't necessary: The OS will do the path lookup for the first argument provided to exec.Command.

File: test/integration/helpers.go [hunk 9]
```diff
@@ -447,3 +489,14 @@ func killProcessFamily(t *testing.T, pid int) {
 		}
 	}
 }
+
+func RunPowershellCmd(args ...string) (string, error) {
```
Comment (tstromberg) [inline test/integration/helpers.go:493]: This appears to be unused.

Comment (medyagh) [reply test/integration/helpers.go:493]: done

File: test/integration/functional_test.go (hunk 3 above)
Comment (medyagh) [reply test/integration/functional_test.go:170]: done

File: test/integration/functional_test.go (hunk 2 above)
Comment (medyagh) [reply test/integration/functional_test.go:162]: done

File: test/integration/helpers.go [hunk 10]
```diff
@@ -111,6 +112,40 @@ func Run(t *testing.T, cmd *exec.Cmd) (*RunResult, error) {
 	return rr, err
 }
 
+// RunInPowershell is a test helper to log a command being executed in powershell \_(ツ)_/¯
```
Comment (tstromberg) [inline test/integration/helpers.go:115]: Please replace this helper with a properly specified *cmd.Cmd structure. It's doing so much opaque and unnecessary work from a testing perspective.

Comment (medyagh) [reply test/integration/helpers.go]: done
